x-pack/plugins/ingest_pipelines/public/application/components/pipeline_editor/components/processor_form/processors/set.tsx

```diff
@@ -101,11 +95,124 @@ const fieldsConfig: FieldsConfig = {
+          id="xpack.ingestPipelines.pipelineEditor.setForm.valueFieldHelpText"
+          defaultMessage="Value for the field"
+        />
+      ),
+      fieldsToValidateOnChange: ['fields.value', 'fields.copy_from'],
+      validations: [
+        {
+          validator: ({ value, path, formData }) => {
+            if (isEmpty(value) && isEmpty(formData['fields.copy_from'])) {
+              return {
+                path,
+                message: i18n.translate('xpack.ingestPipelines.pipelineEditor.requiredField', {
+                  defaultMessage: 'A field value is required.',
+                }),
+              };
+            }
+          },
+        },
+      ],
+    },
+    labelAppend: (
+      <EuiButtonEmpty size="xs" onClick={toggleCustom} data-test-subj="toggleCustomField">
+        {i18n.translate('xpack.ingestPipelines.pipelineEditor.useCopyFromLabel', {
+          defaultMessage: 'Use copy from field',
+        })}
+      </EuiButtonEmpty>
+    ),
+    key: 'value',
+  },
+  copy_from: {
+    path: 'fields.copy_from',
+    euiFieldProps: {
+      'data-test-subj': 'copyFromInput',
+    },
+    config: {
+      type: FIELD_TYPES.TEXT,
+      serializer: from.emptyStringToUndefined,
+      fieldsToValidateOnChange: ['fields.value', 'fields.copy_from'],
+      validations: [
+        {
+          validator: ({ value, path, formData }) => {
+            if (isEmpty(value) && isEmpty(formData['fields.value'])) {
+              return {
+                path,
+                message: i18n.translate('xpack.ingestPipelines.pipelineEditor.requiredField', {
+                  defaultMessage: 'A field value is required.',
+                }),
+              };
+            }
+          },
+        },
+      ],
+      label: i18n.translate('xpack.ingestPipelines.pipelineEditor.setForm.copyFromFieldLabel', {
+        defaultMessage: 'Copy from',
+      }),
+      helpText: (
+        <FormattedMessage
+          id="xpack.ingestPipelines.pipelineEditor.setForm.copyFromFieldHelpText"
+          defaultMessage="Field to copy into {field}."
+          values={{
+            field: <EuiCode>{'Field'}</EuiCode>,
+          }}
+        />
+      ),
+    },
+    labelAppend: (
+      <EuiButtonEmpty size="xs" onClick={toggleCustom} data-test-subj="toggleCustomField">
+        {i18n.translate('xpack.ingestPipelines.pipelineEditor.useValueLabel', {
+          defaultMessage: 'Use value field',
+        })}
+      </EuiButtonEmpty>
+    ),
+    key: 'copy_from',
+  },
+});
+
 /**
  * Disambiguate name from the Set data structure
  */
 export const SetProcessor: FunctionComponent = () => {
-  const [{ fields }] = useFormData({ watch: 'fields.value' });
+  const { getFieldDefaultValue } = useFormContext();
+  const [{ fields }] = useFormData({ watch: ['fields.value', 'fields.copy_from'] });
+
+  const isCopyFromDefined = getFieldDefaultValue('fields.copy_from') !== undefined;
+  const [isCustom, setIsCustom] = useState<boolean>(isCopyFromDefined);
```

Comment:
[nit] `isCustom` seems like possiblly the wrong name here - isn't this representing if the `copy_from` field is enabled or not?